src/main/java/org/openrewrite/java/testing/hamcrest/FlattenAllOf.java

```diff
@@ -0,0 +1,81 @@
+    }
+}
+
+class FlattenAllOfVisitor extends JavaVisitor<ExecutionContext> {
+    private final MethodMatcher ASSERT_THAT_MATCHER = new MethodMatcher("org.hamcrest.MatcherAssert assertThat(..)");
+    private final MethodMatcher ALL_OF_MATCHER = new MethodMatcher("org.hamcrest.Matchers allOf(..)");
+
+    @Override
+    public J visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+        J.MethodDeclaration md = (J.MethodDeclaration) super.visitMethodDeclaration(method, ctx);
+
+        List<Statement> newStatements = new ArrayList<>();
+        for (Statement statement : md.getBody().getStatements()) {
+            if (statement instanceof J.MethodInvocation) {
+                J.MethodInvocation assertThat = (J.MethodInvocation) statement;
+                if (ASSERT_THAT_MATCHER.matches(assertThat)) {
+                    List<Expression> assertThatArguments = assertThat.getArguments();
+                    Expression actual = assertThatArguments.get(assertThatArguments.size() - 2);
+                    Expression matcherExpression = assertThatArguments.get(assertThatArguments.size() - 1);
+                    if (ALL_OF_MATCHER.matches(matcherExpression)) {
+                        J.MethodInvocation allOf = (J.MethodInvocation) matcherExpression;
+                        List<Expression> matchers = allOf.getArguments();
+                        if (1 < matchers.size()) { // Iterable? Inadvertent use of allOf with single Matcher?
+                            for (Expression matcher : matchers) {
+                                JavaTemplate template = JavaTemplate.builder("assertThat(#{any(java.lang.Object)}, #{any(org.hamcrest.Matcher)});")
+                                        .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "hamcrest-2.2"))
+                                        .staticImports("org.hamcrest.MatcherAssert.assertThat")
+                                        .build();
+                                // TODO figure out correct use of cooridnates and cursor when inserting multiple new statements
+                                JavaCoordinates coordinates = newStatements.isEmpty() ?
+                                        assertThat.getCoordinates().replace() :
+                                        newStatements.get(newStatements.size() - 1).getCoordinates().after();
+                                Cursor cursor = null;
+                                J.MethodInvocation newAssertThat = template.apply(cursor, coordinates, actual, matcher);
+                                newStatements.add(newAssertThat);
+                            }
+                            continue;
+                        }
+                    }
+
+                }
+            }
+            // Keep the original statement
+            newStatements.add(statement);
+        }
+
+        return md.withBody(md.getBody().withStatements(newStatements));
```

Comment:
Might want to add a conditional here to only return a different element if there were changes to newStatements.

Reply:
The code should instead be using `ListUtils#map()`. That method automatically figures out if there were any changes to any of the elements and if not returns the original input list. That is nice because it avoids unnecessarily allocating a new list.

Note: In this case the `J.Body#withStatements()` will not be able to simply return `this`, just because the list is still the same instance. It will still end up checking all the elements for referential equality. But if all elements are indeed the same, then the `withStatements()` method will return `this`. So a conditional boolean variable is IMHO unnecessary, if using `ListUtils#map()` here.

Reply:
thanks again; solved differently by reverting to visitMethodInvocation, and returning a block of statements there which we unwrap after visit.